frontend/src/metabase/admin/performance/hooks/useConfirmOnRouteLeave.unit.spec.tsx

```diff
@@ -0,0 +1,107 @@
+import userEvent from "@testing-library/user-event";
+import { useRef, useState } from "react";
+import { Route } from "react-router";
+
+import { act, renderWithProviders, screen } from "__support__/ui";
+import { checkNotNull } from "metabase/lib/types";
+
+import { useConfirmOnRouteLeave } from "./useConfirmOnRouteLeave";
+
+const ConfirmingPage = () => {
+  const [showConfirm, setShowConfirm] = useState(false);
+  const onConfirmRef = useRef<(() => void) | null>(null);
+
+  useConfirmOnRouteLeave({
+    shouldConfirm: true,
+    // Use a simple async-like confirmation UI (not window.confirm)
+    confirm: (onConfirm) => {
+      onConfirmRef.current = onConfirm;
+      setShowConfirm(true);
+    },
+  });
+
+  return (
+    <div>
+      <div>Form B</div>
+      {showConfirm && (
+        <div role="dialog">
+          <div>Are you sure you want to leave?</div>
+          <button
+            onClick={() => {
+              onConfirmRef.current?.();
+              setShowConfirm(false);
+            }}
+          >
+            Yes
+          </button>
+          <button onClick={() => setShowConfirm(false)}>No</button>
+        </div>
+      )}
+    </div>
+  );
+};
+
+const PageA = () => <div>Page A</div>;
+
+describe("useConfirmOnRouteLeave", () => {
+  const setup = () =>
+    renderWithProviders(
+      <div>
+        <Route path="/a" component={PageA} />
+        <Route path="/b" component={ConfirmingPage} />
+      </div>,
+      { withRouter: true, initialRoute: "/a" },
+    );
+
+  // Shared happy-path to the confirmation dialog after attempting to leave /b
+  const navigateToBAndTriggerBack = async (history: any) => {
+    const guardedHistory = checkNotNull(history);
```

Comment:
We usually do the `checkNotNull` assertion in `setup` so that all tests can benefit from it.
With that, `navigateToBAndTriggerBack` wouldn't have to `return` anything.